### ISSUE TITLE
multiplicative_replacement() implementation of delta inconsistent with interpretation

### ISSUE THREAD

**Issue by bhillmann** [biocore/scikit-bio#1551]
The implementation of multiplicative replacement imputes delta in a misleading way. The delta imputed should be proportional to the detection limit, not just the number of features. Otherwise, you could end up with some misleading results of imputation unless if the number of features is in the order of thousands. For your reference, the imputation paper is [here](https://link.springer.com/article/10.1023/A:1023866030544).

For example:

```python
from sckbio.stat.composition import multiplicative_replacement, closure

# This also works on count data directly
def multiplicative_replacement2(mat, delta=None):
    r"""Replace all zeros with small non-zero values
    It uses the multiplicative replacement strategy [1]_ ,
    replacing zeros with a small positive :math:`\delta`
    and ensuring that the compositions still add up to 1.
    Parameters
    ----------
    mat: array_like
       a matrix of proportions where
       rows = compositions and
       columns = components
    delta: float, optional
       a small number to be used to replace zeros
       If delta is not specified, then the default delta is
       :math:`\delta = \frac{1}{N^2}` where :math:`N`
       is the number of components
    Returns
    -------
    numpy.ndarray, np.float64
       A matrix of proportions where all of the values
       are nonzero and each composition (row) adds up to 1
    Raises
    ------
    ValueError
       Raises an error if negative proportions are created due to a large
       `delta`.
    Notes
    -----
    This method will result in negative proportions if a large delta is chosen.
    References
    ----------
    .. [1] J. A. Martin-Fernandez. "Dealing With Zeros and Missing Values in
           Compositional Data Sets Using Nonparametric Imputation"
    Examples
    --------
    >>> import numpy as np
    >>> from skbio.stats.composition import multiplicative_replacement
    >>> X = np.array([[.2,.4,.4, 0],[0,.5,.5,0]])
    >>> multiplicative_replacement(X)
    array([[ 0.1875,  0.375 ,  0.375 ,  0.0625],
           [ 0.0625,  0.4375,  0.4375,  0.0625]])
    """
    mat = np.array(mat)
    z_mat = (mat == 0)

    num_feats = mat.shape[-1]
    tot = np.sum(z_mat, axis=-1, keepdims=True)

    if delta is None:
        delta = np.min((np.min(mat[mat > 0])*.55, (1. / num_feats)**2))

    zcnts = 1 - ((tot * delta)/mat.sum(axis=-1, keepdims=True))
    if np.any(zcnts) < 0:
        raise ValueError('The multiplicative replacment created negative '
                         'proportions. Consider using a smaller `delta`.')
    mat = np.where(z_mat, delta, zcnts * mat)
    return mat.squeeze()

X = closure(np.array([0, 1, 2, 4, 8, 16, 32, 64, 128]))

X

#array([ 0.        ,  0.00392157,  0.00784314,  0.01568627,  0.03137255, 0.0627451 ,  0.1254902 ,  0.25098039,  0.50196078])

multiplicative_replacement(X)

#array([ 0.01234568,  0.00387315,  0.00774631,  0.01549262,  0.03098523, 0.06197047,  0.12394093,  0.24788187,  0.49576374])

multiplicative_replacement2(X)

#array([ 0.00215686,  0.00391311,  0.00782622,  0.01565244,  0.03130488, 0.06260977,  0.12521953,  0.25043906,  0.50087812])
```

**Comment by mortonjt** [comment]
Right - multiplicative imputation is generally awkward in this realm.  Typically we are dealing with counts, where there is no detection limit (which is specified by `delta`).  The default `delta` is not ideal when you have a smaller number of features, which is why there is the option to impute based on the user choice of `delta`.

I'm not entirely sure about the merits of adding in this sort of functionality in -- it's not clear me what an appropriate detection limit would be.  We also know that there are number of much better imputation methods (see this [paper](https://academic.oup.com/bioinformatics/article/34/3/372/4157442)).

**Comment by bhillmann** [comment]
The implementation that I provided here ```multiplicative_replacement2()``` works on actual count data, it closes to the sum of the composition vector specified. In the implementation previously, you assume a closure to 1, so it actually only works only on relative abundance data, or sample vectors that sum to 1.

Also, delta is the value that you replace 0 values with. In the version currently implemented, the delta can be set to a value that is greater than the smallest value in your matrix, an estimation of the detection limit. I specify the detection limit to be the minimum of the input matrix, and then multiply that value simply by .55 which, is what was recommended in the original paper linked in the description. Furthermore, I kept the imputation of delta that was previously provided, and took the minimum of the two imputed deltas to be on the safe side.

Simply put, if my count vector is ```x = [0, 1, 2, 4, 8, 16, 32, 64, 128]```, I shouldn't replace 0 with a value that is greater than 1. This can happen in the original implementation, as shown in the example.

The new approach that you provided is interesting, but it requires dense longitudinal data for subjects, and also requires Bayesian prior assumptions. This doesn't work in cross-sectional studies, and is difficult to define a dense sampling.
